File: CivitasRuby/pkg/CivitasRuby-0.0.1/lib/civitas_juego.rb
#encoding:utf-8

require_relative 'jugador'
require_relative 'gestor_estados'
require_relative 'estados_juego'
require_relative 'dado'
require_relative 'sorpresa'
require_relative 'tablero'
require_relative 'titulo_propiedad'
require_relative 'diario'

module Civitas
  
  class CivitasJuego
    
    
    def initialize(nombres)
      @jugadores = Array.new
      for nombre in nombres
        jugador = Jugador.new_por_nombre(nombre)
        @jugadores << jugador
      end 
      
      @indice_jugador_actual = Dado.instance.quienEmpieza(@jugadores.length)
      @gestor_estados = Gestor_estados.new
      @estado = @gestor_estados.estado_inicial
      @tablero = Tablero.new(Casilla.Carcel)
      @mazo = MazoSorpresas.new
      
      inicializar_mazo_sorpresas(@tablero) 
      inicializar_tablero(@mazo)
     
      
      
    end
    
    
    def muestra_tablero
      @tablero.muestra_tablero
    end
    
    def cancelar_hipoteca(ip)
      
      @jugadores[@indice_jugador_actual].cancelar_hipoteca(ip)
      
    end
    
    
    
    def comprar
      
      res=false
      jugador_actual=@jugadores[@indice_jugador_actual]
      num_casilla_actual=jugador_actual.num_casilla_actual
      casilla=@tablero.casillas[num_casilla_actual]
      titulo=casilla.titulo
      res=jugador_actual.comprar(titulo)
      
      res
      
    end
    
    def construir_casa(ip)
      @jugadores[@indice_jugador_actual].construir_casa(ip)
    end
    
    def construir_hotel(ip)
      @jugadores[@indice_jugador_actual].construir_hotel(ip)
    end
    
    def actualizar_info
      
      @jugadores[@indice_jugador_actual].to_string
      
      
      
    end
    
    def final_del_juego
      algunArruinado = false;
      for jugador in @jugadores
        if(jugador.en_bancarrota)
          algunArruinado = true
        end
      end
      
      algunArruinado
      
    end
    
    def get_casilla_actual
      @tablero.casilla(@jugadores[@indice_jugador_actual].num_casilla_actual)
    end
    
    def get_jugador_actual
      @jugadores[@indice_jugador_actual]
    end
    
    def hipotecar(ip)
      get_jugador_actual.hipotecar(ip)
    end
    
    def info_jugador
      get_jugador_actual.to_string
    end
    
    def salir_carcel_pagando
      @jugadores[@indice_jugador_actual].salirCarcelPagando();
    end
    
    def salir_carcel_tirando
      @jugadores[@indice_jugador_actual].salirCarcelTirando();
    end
    
    def siguiente_paso
      jugador_actual = @jugadores[@indice_jugador_actual]
      operacion=@gestor_estados.operaciones_permitidas(jugador_actual, @estado)
      
      if(operacion==OperacionesJuego::PASAR_TURNO)
         pasar_turno
         siguiente_paso_completado(operacion)
         
       elsif(operacion==OperacionesJuego::AVANZAR)
           avanza_jugador
           siguiente_paso_completado(operacion)
         end
         operacion
      end
    
    
    def siguiente_paso_completado(operacion)
      
      @estado=@gestor_estados.siguiente_estado(get_jugador_actual, @estado, operacion)
      
    end
    
    def vender(ip)
      get_jugador_actual.vender(ip)
    end
    
    def avanza_jugador
      
      jugador_actual = @jugadores[@indice_jugador_actual]
      posicion_actual = jugador_actual.num_casilla_actual
      tirada = Dado.instance.tirar
      posicion_nueva = @tablero.nueva_posicion(posicion_actual,tirada)
      
      casilla = @tablero.casilla(posicion_nueva)
      
      contabilizar_pasos_por_salida(jugador_actual)
      
      jugador_actual.mover_a_casilla(posicion_nueva)
      casilla.recibe_jugador(@indice_jugador_actual,@jugadores)
      
      contabilizar_pasos_por_salida(jugador_actual)
       
    end
   

    private
    
    def contabilizar_pasos_por_salida(jugador_actual)

      pasos = @tablero.get_por_salida
      while pasos > 0
        
        jugador_actual.pasa_por_salida
        pasos = @tablero.get_por_salida
        
      end
      
    end
    
    def inicializar_mazo_sorpresas(tablero)

      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::IRCARCEL,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::IRCASILLA,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::PAGARCOBRAR,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::PORCASAHOTEL,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::PORJUGADOR,tablero))
      @mazo.alMazo(Sorpresa.new_por_mazo(TipoSorpresa::SALIRCARCEL,tablero))

    end
    
    def inicializar_tablero(mazo)
      casillas = ["Calle Mulillas","Calle Dtor Oloriz" , "Calle Froilán"]
      
      for casilla in casillas
        titulo = TituloPropiedad.new(casilla,200,1.25,500,24000,100)
        @tablero.añadeCasilla(Casilla.new_por_titulo(titulo))
      end
      
      
    end
    
    def pasar_turno
      if @indice_jugador_actual < @jugadores.length - 1
        @indice_jugador_actual += 1
      else
        @indice_jugador_actual = 0
      end
      
    end
    
    def ranking
      
      @jugadores.sort! { |a,b| a.saldo <=> b.saldo }
      
    end
    
  end
end

File: CivitasJava/src/civitas/CivitasJuego.java
package civitas;


import java.util.Collections;
import java.util.ArrayList;
import java.util.Random;
public class CivitasJuego {
    
    private int indiceJugadoractual;
    private ArrayList<Jugador> jugadores;
    private MazoSorpresas mazo;
    private Tablero tablero;
    private GestorEstados gestorEstados; 
    private EstadosJuego estado;
    
    CivitasJuego(ArrayList<String> nombres)
    {
       
       jugadores = new ArrayList<Jugador>();
       
       for(String n:nombres)
            jugadores.add(new Jugador(n));
       
       gestorEstados = new GestorEstados();
       estado = gestorEstados.estadoInicial();
       
       indiceJugadoractual = Dado.getInstance().quienEmpieza(jugadores.size());
       
       mazo = new MazoSorpresas();

        inicializaTablero(mazo);
        inicializarMazoSorpresas(tablero);

       
    }

    public int getIndiceJugadoractual() {
        return indiceJugadoractual;
    }

    public MazoSorpresas getMazo() {
        return mazo;
    }
    
    void actualizarInfo(){
        
        Jugador jugador = jugadores.get(indiceJugadoractual);
        
        System.out.println(jugador.toString());
        
        
    }
//    
    private void avanzaJugador()
    {
       Jugador jugadorActual=jugadores.get(indiceJugadoractual);
       int posicionActual=jugadorActual.getNumCasillaActual();
       int tirada=Dado.getInstance().tirar();
       int posicionNueva=tablero.nuevaPosicion(posicionActual,tirada);
       Casilla casilla=tablero.getCasilla(posicionNueva);
       
       
       contabilizarPasosPorSalida(jugadorActual);
       jugadorActual.moverACasilla(posicionNueva);
       casilla.recibeJugador(indiceJugadoractual, jugadores);
       contabilizarPasosPorSalida(jugadorActual);
       
    }
    
    public Boolean cancelarHipoteca(int ip)
    {
        return jugadores.get(indiceJugadoractual).cancelarHipoteca(ip);
    }        
    public Boolean comprar()
    {   
        boolean res=false;
        Jugador jugadorActual=jugadores.get(indiceJugadoractual);
        int numCasillaActual=jugadorActual.getNumCasillaActual();
        Casilla casilla=tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo=casilla.getTituloPropiedad();
        res=jugadorActual.comprar(titulo);
        
        return res;
    }
    
    public Boolean construirCasa(int ip){
       return jugadores.get(indiceJugadoractual).construirCasa(ip);
    }
    
    public Boolean construirHotel(int ip)
    {
       return jugadores.get(indiceJugadoractual).construirHotel(ip);        
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual)
    {
        while(tablero.getPorSalida()>0)
        {
            jugadorActual.pasaPorSalida();
        }   
    }
    
    public Boolean finalDelJuego()
    {
        Boolean resultado=false;
        for(int i=0;i<jugadores.size();i++)
        {
            resultado=jugadores.get(i).enBancarrota();
        }
        
        return resultado;       
    }
    
    public Casilla getCasillaActual()
    {
        return tablero.getCasilla(jugadores.get(indiceJugadoractual).getNumCasillaActual());        
    }
    
    public Jugador getJugadorActual()
    {
        return jugadores.get(indiceJugadoractual);
    } 
    
    public Boolean hipotecar(int ip)
    {
       return jugadores.get(indiceJugadoractual).hipotecar(ip);
    }            
    
    public String infoJugadorTexto()
    {
        return jugadores.get(indiceJugadoractual).toString();
    }   
    
    private void inicializarMazoSorpresas(Tablero tablero)
    {
        
        //Creación de una sorpresa de cada tipo
        //Almacenamiento en el mazo
        Random ram=new Random();
        for(TipoSorpresa tipo:TipoSorpresa.values())
        {
            switch(tipo)
            {
                case IRCARCEL:
                    mazo.alMazo(new Sorpresa(tipo,tablero));   
                    break;
                case IRCASILLA:
                   mazo.alMazo(new Sorpresa(tipo,tablero,ram.nextInt(tablero.size()),tipo.toString()));
                   break;
                case SALIRCARCEL:
                    mazo.alMazo(new Sorpresa(tipo,mazo));
                    break;
                default:
                    mazo.alMazo(new Sorpresa(tipo,ram.nextInt(400)+100,tipo.toString()));
            }
        }
        
    } 
    
    private void inicializaTablero(MazoSorpresas mazo)
    {
       tablero=new Tablero(Casilla.getCarcel());
       for(int i=1;i<3;i++)
       {
        TituloPropiedad prop=new TituloPropiedad("Calle"+i,100*i,3,50*i,500+100*i,200);
        Casilla casilla=new Casilla(prop);
        tablero.añadeCasilla(casilla); 
       }
       Casilla cas1;

       cas1=new Casilla(mazo,"SorpreCasilla");
       tablero.añadeCasilla(cas1);
       tablero.añadeCasilla(new Casilla(100f,"Impuesto"));
       tablero.añadeJuez();
       

       
    }
    
    private void pasarTurno()
    {   
        if(indiceJugadoractual<jugadores.size()-1){
            indiceJugadoractual++;     
        }else{
            indiceJugadoractual=0;
        }
        
    }
    
    private ArrayList<Jugador> ranking()
    {
      
      Collections.sort(jugadores);
      return jugadores;
    }
    
    
    public Boolean salirCarcelPagando()
    {
       return jugadores.get(indiceJugadoractual).salirCarcelPagando(); 
    }
    
    public Boolean salirCarcelTirando()
    {
       return jugadores.get(indiceJugadoractual).salirCarcelTirando();        
    }
    
    public OperacionesJuego siguientePaso()
    {
       Jugador jugadorActual=jugadores.get(indiceJugadoractual);
       OperacionesJuego operacion=gestorEstados.operacionesPermitidas(jugadorActual, estado);
       
       if(operacion==OperacionesJuego.PASAR_TURNO)
       {
           pasarTurno();
       }
       else if(operacion==OperacionesJuego.AVANZAR)
       {
           avanzaJugador();
       }
            siguientePasoCompletado(operacion);
       
       return operacion;
       
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion)
    {

        estado=gestorEstados.siguienteEstado(getJugadorActual(),estado,operacion);
       
    }
    
    public Boolean vender(int ip)
    {
       return jugadores.get(indiceJugadoractual).vender(ip);
    }
    
    
    
    public static void main(String args[]){
        
        ArrayList jugadores = new ArrayList<String>();
        
        jugadores.add("paco");
        jugadores.add("juan");
        
        CivitasJuego game = new CivitasJuego(jugadores);
        
        game.actualizarInfo();
        
        MazoSorpresas generado = game.getMazo();
        
        Sorpresa s = generado.siguiente();
        
        
        
    }
    
    public String getEstado()
    {
        return estado.toString();
    }
    
}
